File: samples/dnn/vlm_ocr.py
# This file is part of OpenCV project.
# It is subject to the license terms in the LICENSE file found in the top-level directory
# of this distribution and at http://opencv.org/license.html.
# Copyright (C) 2026, BigVision LLC, all rights reserved.
# Third party copyrights are property of their respective owners.

'''
This is a sample script demonstrating cv2.vlm: a single API for running vision-language
OCR / document-understanding inference either locally (PaddleOCR-VL-1.5, Granite-Docling-258M)
or via a hosted cloud API (OpenAI, Anthropic, Gemini, Grok), given a model type, a local ONNX
export directory (or cloud model name), and an input image.

Run the script:

    python vlm_ocr.py --model_type=paddleocr-vl --model_dir=<dir> --input=<path-to-image>
    python vlm_ocr.py --model_type=openai --api_key=<key> --input=<path-to-image>
'''

import argparse
import cv2 as cv

MODEL_TYPES = {
    'paddleocr-vl': cv.vlm.VLM_MODEL_PADDLEOCR_VL,
    'granite-docling': cv.vlm.VLM_MODEL_GRANITE_DOCLING,
    'openai': cv.vlm.VLM_MODEL_OPENAI,
    'anthropic': cv.vlm.VLM_MODEL_ANTHROPIC,
    'gemini': cv.vlm.VLM_MODEL_GEMINI,
    'grok': cv.vlm.VLM_MODEL_GROK,
}

CLOUD_MODEL_TYPES = {'openai', 'anthropic', 'gemini', 'grok'}

def parse_args():
    parser = argparse.ArgumentParser(description='Use this script to run vision-language OCR / '
                                                  'document-understanding inference in OpenCV',
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('--model_type', type=str, required=True, choices=sorted(MODEL_TYPES),
                        help='Which VLM to run.')
    parser.add_argument('--model_dir', type=str, default='',
                        help='Local model types: path to the ONNX export directory. '
                             'Cloud model types: provider model name, e.g. gpt-4o (required, no default).')
    parser.add_argument('--api_key', type=str, default='',
                        help='API key for cloud model types (openai/anthropic/gemini/grok); ignored otherwise.')
    parser.add_argument('--input', '-i', type=str, required=True, help='Path to the input image.')
    parser.add_argument('--prompt', type=str, default='', help="Task prompt (default: the model's built-in prompt).")
    parser.add_argument('--max_new_tokens', type=int, default=512, help='Maximum number of new tokens to generate.')
    parser.add_argument('--engine', type=str, default='new', choices=['new', 'ort'],
                        help='Local model types only: dnn engine used to load each ONNX sub-model.')
    parser.add_argument('--device', type=str, default='cpu', choices=['cpu', 'cuda'],
                        help='Local model types only: compute device.')
    args = parser.parse_args()

    if not args.model_dir:
        parser.error('--model_dir is required (local: ONNX export directory, cloud: provider model name)')
    if args.model_type in CLOUD_MODEL_TYPES and not args.api_key:
        parser.error('--api_key is required for cloud model types')
    return args

if __name__ == '__main__':

    args = parse_args()

    device = 'cloud' if args.model_type in CLOUD_MODEL_TYPES else args.device

    print(f'Preparing {args.model_type} model...')
    model = cv.vlm.create(MODEL_TYPES[args.model_type], args.model_dir,
                          args.engine, device, args.api_key)

    print(f'Running inference on {args.input}...')
    results = model.inferDocument(args.input, args.prompt, args.max_new_tokens)
    for i, text in enumerate(results):
        print(f'Page {i + 1}:\n{text}')

File: samples/dnn/vlm_ocr.cpp
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html.
// Copyright (C) 2026, BigVision LLC, all rights reserved.
// Third party copyrights are property of their respective owners.

/*
 * This is a sample program demonstrating cv::vlm::create(): a single API for running
 * vision-language OCR / document-understanding inference either locally (PaddleOCR-VL-1.5,
 * Granite-Docling-258M) or via a hosted cloud API (OpenAI, Anthropic, Gemini, Grok), given a
 * model type, a local ONNX export directory (or cloud model name), and an input image.
 *
 * Run the sample:
 *
 *      ./vlm_ocr --model_type=paddleocr-vl --model_dir=<dir> --input=<path-to-image>
 *      export OPENAI_API_KEY=...   # or ANTHROPIC_API_KEY / GEMINI_API_KEY / XAI_API_KEY
 *      ./vlm_ocr --model_type=openai --model_dir=<provider model name> --input=<path-to-image>
 *
 * Cloud model types read the key from the provider's usual environment variable. Prefer that
 * over --api_key, which puts the key in your shell history and in the process list.
 */

#include <iostream>
#include <map>

#include <opencv2/core.hpp>
#include <opencv2/dnn.hpp>
#include <opencv2/vlm.hpp>

using namespace cv;
using namespace cv::vlm;
using namespace std;

int main(int argc, char** argv)
{
    const string keys =
        "{ help h         |      | Print help message }"
        "{ model_type     |      | Which VLM to run: paddleocr-vl, granite-docling, openai, anthropic, gemini or grok }"
        "{ model_dir      |      | Local model types: path to the ONNX export directory. "
        "Cloud model types: provider model name, taken from the provider's model list (required, no default) }"
        "{ api_key        |      | API key for cloud model types; ignored otherwise. Leave unset to read it from "
        "OPENAI_API_KEY / ANTHROPIC_API_KEY / GEMINI_API_KEY / XAI_API_KEY instead }"
        "{ input i        |      | Path to the input image (.png/.jpg/.jpeg) }"
        "{ prompt         |      | Task prompt (defaults to the model's built-in prompt) }"
        "{ max_new_tokens | 512  | Maximum number of new tokens to generate }"
        "{ engine         | opencv | Local model types only: dnn engine used to load each ONNX sub-model: opencv }"
        "{ device         | cpu  | Local model types only: compute device: cpu or cuda }";

    CommandLineParser parser(argc, argv, keys);
    parser.about("Use this sample to run vision-language OCR / document-understanding inference in OpenCV");
    if (parser.has("help") || !parser.has("model_type") || !parser.has("input"))
    {
        parser.printMessage();
        return 0;
    }

    static const map<string, VLMModelType> MODEL_TYPES = {
        {"paddleocr-vl",     VLM_MODEL_PADDLEOCR_VL},
        {"granite-docling",  VLM_MODEL_GRANITE_DOCLING},
        {"openai",           VLM_MODEL_OPENAI},
        {"anthropic",        VLM_MODEL_ANTHROPIC},
        {"gemini",           VLM_MODEL_GEMINI},
        {"grok",             VLM_MODEL_GROK},
    };

    string modelTypeArg = parser.get<String>("model_type");
    auto it = MODEL_TYPES.find(modelTypeArg);
    if (it == MODEL_TYPES.end())
    {
        cerr << "Unknown model_type: " << modelTypeArg << endl;
        return 1;
    }
    VLMModelType modelType = it->second;
    bool isCloud = (modelType == VLM_MODEL_OPENAI || modelType == VLM_MODEL_ANTHROPIC ||
                    modelType == VLM_MODEL_GEMINI || modelType == VLM_MODEL_GROK);

    if (!parser.has("model_dir"))
    {
        cerr << "model_dir is required: local model types need the ONNX export directory, "
                "cloud model types need the provider's model name" << endl;
        return 1;
    }

    string engine = parser.get<String>("engine");
    if (!isCloud && engine != "opencv")
    {
        cerr << "Unknown engine: " << engine << " (expected opencv)" << endl;
        return 1;
    }

    string device = "cloud";
    if (!isCloud)
    {
        device = parser.get<String>("device");
        if (device != "cpu" && device != "cuda")
        {
            cerr << "Unknown device: " << device << " (expected cpu or cuda)" << endl;
            return 1;
        }
    }

    string modelDir = parser.get<String>("model_dir");
    string apiKey = parser.get<String>("api_key");
    string inputPath = parser.get<String>("input");
    string prompt = parser.get<String>("prompt");
    int maxNewTokens = parser.get<int>("max_new_tokens");

    cout << "Preparing " << modelTypeArg << " model..." << endl;
    Ptr<VLMModel> model = create(modelType, modelDir, engine, device, apiKey);

    cout << "Running inference on " << inputPath << "..." << endl;
    vector<String> results = model->inferDocument(inputPath, prompt, maxNewTokens);
    for (size_t i = 0; i < results.size(); i++)
        cout << "Page " << (i + 1) << ":\n" << results[i] << endl;

    return 0;
}
